### ISSUE TITLE
TUManager.cpp

### ISSUE THREAD

**Issue by 3246251196** [Sarcasm/irony-mode#282]
@Sarcasm 
Hi, is there any reason why in ::parse if a reference to a cached TU is returned (flags have not changed or anything) that the re-parse function gets called regardless? Can we not just return the TU ?

Cheers.

**Comment by Sarcasm** [comment]
> Hi, is there any reason why in ::parse if a reference to a cached TU is returned (flags have not changed or anything) that the re-parse function gets called regardless? Can we not just return the TU ?

Maybe the function is badly named.
Right now there are 2 ways the buffers are analyzed by irony-mode.
Either you do completion or you want to [re]parse the buffer to check for new diagnostics. parse() is called when the buffer content has changed, we need to reparse to be up to date. If we were just returning the TU we would be working on outdated content.

> I ask only because on large projects clang_parseTranslationUnit/clang_reparseTranslationUnit can take 8seconds each.

8 seconds is a lot but that depends on the project, heavily templated code using boost is definitely not the fastest.

**Comment by 3246251196** [comment]
@Sarcasm

Well it takes a while with the initial parse. Then subsequent completions are fairly quick. It's not like I am waiting ffor that long on every completion.  But this is why I created the function hook to automatically start parsing when you enter a file. In the time it takes for the initial parse you are still probably browsing the file and looking at source code. It means by the time you do a completion things are fairly quick.


To help large projects reduce time would it be possible to actually use an outdated tu reference and get those completions.

So for example I open a buffer. It all parses. I complete a symbol.  All fine. I add in a header file and so other stuff which out dates things. But when I complete the same symbol again irony can use the outdated tu ref while in the background updated it and perhaps appending to the completion list anything it sees fit. This way completions a quicker and still being updated. 

Generally I have found that when things do get outdated you are still searching for a symbol that has not changed even in the updated tu.  That is generally the case.

Maybe I can work on hacking how tumanager does this?  And adding an option to irony ffor large projects where this would be useful. 

Iron-server --re-use-outdated-tu-for-large-projects

You know, something to that effect.

Cheers.

**Comment by Sarcasm** [comment]
> To help large projects reduce time would it be possible to actually use an outdated tu reference and get those completions.

Yeah I think there is a serialization API in Clang. It would make sense to use it, even to just reduce memory usage (which to my suprise nobody complained about yet :)).

> Iron-server --re-use-outdated-tu-for-large-projects

I think it could be the default, no need for an extra option as it is supposed to be transparent/correct. What we could do is just let the user specify the destination dir, the quantity of files to keep in cache or things like that.
